### PR TITLE
[chat] refactor trainer class

### LINKED ISSUE #4077: [chat] refactor trainer class
## Description

The base `Trainer` class contains methods like `_on_fit_start`and `_on_fit_end`, which is only used in RL training.

## TODOs
- [x] Separate current base `Trainer` into `SLTrainer` and `RLTrainer`.
- [x] Update `SFTTrainer`.
- [x] Update `RMTrainer`.
- [x] Update `PPOTrainer`.

### PR DESCRIPTION
**Pull request by CWHer**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

Close #4077.

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: applications/Chat/coati/trainer/base.py [hunk 1]
```diff
@@ -1,54 +1,109 @@
...
     def __init__(self,
                  strategy: Strategy,
-                 max_epochs: int = 1,
-                 dataloader_pin_memory: bool = True,
-                 callbacks: List[Callback] = [],
-                 **generate_kwargs) -> None:
+                 max_epochs: int,
+                 model: nn.Module,
+                 optimizer: Optimizer,
+                 train_dataloader: DataLoader,
```
Comment (ver217) [inline applications/Chat/coati/trainer/base.py:34]: Why pass data loader in `__init__()`? This is usually passed in `fit()`.

File: applications/Chat/coati/trainer/ppo.py [hunk 2]
```diff
@@ -99,76 +118,20 @@ def __init__(self,
...
-            dataloader = self.strategy.setup_dataloader(self.replay_buffer, self.dataloader_pin_memory)
-        if self.sample_replay_buffer:
-            pbar = tqdm(range(self.max_epochs), desc='Train epoch', disable=not is_rank_0())
-            for _ in pbar:
-                experience = self.replay_buffer.sample()
-                experience.to_device(self.device)
-                metrics = self.training_step(experience)
-                pbar.set_postfix(metrics)
+    def _make_experience(self, collect_step: int) -> Experience:
+        prompts = next(iter(self.prompt_dataloader))  # sample a prompt
```
Comment (ver217) [inline applications/Chat/coati/trainer/ppo.py:122]: This usage is wrong and you should fix it.

File: applications/Chat/coati/trainer/sft.py [hunk 3]
```diff
@@ -38,94 +36,93 @@ def __init__(
...
+        for batch_id, batch in enumerate(self.train_dataloader):
+
+            batch = to_device(batch, torch.cuda.current_device())
+            outputs = self.model(batch["input_ids"],
+                                 attention_mask=batch["attention_mask"],
+                                 labels=batch["labels"])
+
+            loss = outputs.loss
+
+            if loss >= 2.5 and is_rank_0():
```
Comment (ver217) [inline applications/Chat/coati/trainer/sft.py:69]: This is not reasonable. I think we can remove this.

File: applications/Chat/coati/trainer/ppo.py (hunk 2 above)
Comment (CWHer) [reply applications/Chat/coati/trainer/ppo.py:122]: The original version wrote `prompts = next(iter(self.prompt_dataloader))`. [Link](https://github.com/hpcaitech/ColossalAI/blob/main/applications/Chat/coati/trainer/ppo.py) at line 153.

Comment (CWHer) [reply applications/Chat/coati/trainer/ppo.py:122]: Isn't this written on purpose? 

https://github.com/hpcaitech/ColossalAI/blob/95e95b6d588d27aeba8d62e6656392f771b69aee/applications/Chat/coati/trainer/ppo.py#L183-L186

The usage of `pretrain_dataloader` also follows this pattern.

File: applications/Chat/coati/trainer/sft.py (hunk 3 above)
Comment (CWHer) [reply applications/Chat/coati/trainer/sft.py]: This is fixed in 9568185.

File: applications/Chat/coati/trainer/base.py (hunk 1 above)
Comment (CWHer) [reply applications/Chat/coati/trainer/base.py]: This is fixed in 211dfb5 and 69ab71b.

File: applications/Chat/coati/trainer/ppo.py (hunk 2 above)
Comment (CWHer) [reply applications/Chat/coati/trainer/ppo.py]: This is fixed in 128ec4f.

**Review by ver217 — APPROVED** [review]

**Pull request merged**